src/JsonLd/Serializer/ItemNormalizer.php

```diff
@@ -64,12 +72,16 @@ public function normalize($object, $format = null, array $context = [])
         $resourceMetadata = $this->resourceMetadataFactory->create($resourceClass);
         $data = $this->addJsonLdContext($this->contextBuilder, $resourceClass, $context);
 
+        $context = $this->addJsonLdDocumentContext($object, $context);
+
         $rawData = parent::normalize($object, $format, $context);
         if (!is_array($rawData)) {
             return $rawData;
         }
 
-        $data['@id'] = $this->iriConverter->getIriFromItem($object);
+        $identifiers = $this->itemIdentifiersExtractor->getIdentifiersFromItem($object);
+
+        $data['@id'] = count(array_filter($identifiers)) > 0 ? $this->iriConverter->getIriFromItem($relatedObject) : $this->blankNodeIdentifiersGenerator->getBlankNodeIdentifier($relatedObject, $context['jsonld_document_root']);
```

Comment:
> Shouldn't we add a method that takes identifiers and resolves to one of those to avoid repetition?

Sure. I'll move it to a private method.

> Is $context['jsonld_document_root'] used only to cache the spl_object_hash?

We need to keep track of which JSON-LD document we're in. But of course there are many ways of doing it. I agree this is not ideal.